File: leetcode/enumerated/cpp/40.cpp
// https://leetcode.com/problems/combination-sum-ii/
#include <algorithm>
#include <vector>

using std::vector;

class Solution {
public:
    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        vector<vector<int>> result;
        sort(candidates.begin(), candidates.end());

        while (!candidates.empty() && candidates.back() > target) {
            candidates.pop_back();
        }

        vector<int> current_combination;
        backtrack(0, candidates, target, 0, current_combination, result);

        return result;
    }

private:
    void backtrack(int start, const vector<int>& candidates, int target,
                    int current_sum, vector<int>& current_combination,
                    vector<vector<int>>& result) {
        if (current_sum == target) {
            result.push_back(current_combination);
            return;
        }

        if (current_sum > target) {
            return;
        }

        for (int i = start; i < candidates.size(); ++i) {
            if (i > start && candidates[i] == candidates[i - 1]) {
                continue;
            }

            current_combination.push_back(candidates[i]);
            backtrack(i + 1, candidates, target, current_sum + candidates[i], current_combination, result);
            current_combination.pop_back();
        }
    }
};

File: leetcode/enumerated/py/40.py
# https://leetcode.com/problems/combination-sum-ii/

class Solution:
    def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:
        candidates.sort()
        result = []

        while len(candidates) > 0 and candidates[-1] > target:
            candidates.pop()

        def backtrack(start, current_combination, current_sum):
            if current_sum == target:
                result.append(list(current_combination))
                return

            if current_sum > target:
                return

            for i in range(start, len(candidates)):
                if i > start and candidates[i] == candidates[i - 1]:
                    continue

                current_combination.append(candidates[i])
                backtrack(i + 1, current_combination, current_sum + candidates[i])
                current_combination.pop()

        backtrack(0, [], 0)
        return result
